lib/Monad_Commute.thy

```diff
@@ -0,0 +1,195 @@
+(*
+ * Copyright 2023, Proofcraft Pty Ltd
+ * Copyright 2020, Data61, CSIRO (ABN 41 687 119 230)
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ *)
+
+(* The monad_commute predicate + lemmas. *)
+
+theory Monad_Commute
+  imports
+    Monad_Equations
+    Monad_Lists (* for mapM_x *)
+begin
+
+
+definition monad_commute where
+  "monad_commute P a b \<equiv>
+   \<forall>s. (P s \<longrightarrow> ((do x\<leftarrow>a;y\<leftarrow>b; return (x, y) od) s) = ((do y\<leftarrow>b;x\<leftarrow>a; return (x, y) od) s))"
+
+
+lemma monad_eq:
+  "a s = b s \<Longrightarrow>  (a >>= g) s = (b >>= g) s"
+  by (auto simp:bind_def)
+
+lemma monad_commute_simple:
+  "\<lbrakk> monad_commute P a b; P s \<rbrakk> \<Longrightarrow> (do x\<leftarrow>a;y\<leftarrow>b; g x y od) s = (do y\<leftarrow>b;x\<leftarrow>a; g x y od) s"
+  apply (clarsimp simp:monad_commute_def)
```

Comment:
if we care, this file has a lot of `[a-z]:[a-z]` that could use an extra space after `:`

Reply:
since I've been messing with the diff already for `[monad_eq]`, I think I'll apply a search&replace for this one as well.